README.md
# Emoji Keyboard Export

This is the reposition for the Emoji Keyboard Export mod for Factorio.
This mod will export a plugin for the [Emoji Keyboard](https://github.com/gilleswaeber/emoji-keyboard) app for the active Factorio game, including mods, using the active locale.

The plugin consists of a JSON file describing all items, fluids, recipes, technologies, and virtual signals that can be typed using [Rich Text](https://wiki.factorio.com/Rich_text), categorized and with their tag, and a sprite map.

This mod won't do anything when not run in Instrument Mode. [Learn more](#why-instrument-mode).

This is based on [FactorioLab Export](https://github.com/factoriolab/factoriolab-export) by ilbJanissary, available under MIT License. Thank you!

## Getting started

1. Install this mod alongside any mods that should be included in the data set
2. Set up Steam to run this mod in [Instrument Mode](https://lua-api.factorio.com/latest/Instrument.html)
   - In the library, right click Factorio and in Launch Options enter: `--instrument-mod emoji-keyboard-export`
   - Alternately, run Factorio from the command line with these options
   - [Why instrument mode?](#why-instrument-mode)
3. Start Factorio and start a new game, a message will appear indicating the export started (if not, check that this mod and flib mod are enabled, and you used instrument mode)
4. Wait for the export to complete, another message appears in the chat, you can now exit Factorio
5. Copy the contents of `%APPDATA%\Factorio\script-output\emoji-keyboard-export` into the plugin directory of Emoji Keyboard and restart it.

## Why instrument mode?

Running in instrument mode is technically optional but the mod disables itself when it's not used. The source mode, FactorioLab Export can run without it.

The FactorioLab Export code uses instrument mode to check icon data for cases where icons will be rendered at unexpected sizes in the game, to ensure that the generated sprite sheet sizes the icons appropriately. In certain cases icons can bleed out of the expected size or can be scaled in unexpected ways that can only be detected in the data stage. The mod runs these checks in the `instrument-after-data.lua` file to ensure that it has the most up to date icon information from all loaded mods.

Unfortunately, the Factorio Lua runtime exposes no method to determine what size a sprite will be rendered to when using `LuaRendering.draw_sprite`. By checking the icons in the data stage, the mod can predict the size and use the `scale` parameter to ensure all the icons are the desired size.

## Warnings and errors

The code that FactorioLab Export uses to compute recipes has been stripped out, so no known source of error remains.
Please file an issue if anything weird happens.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 2023-07-22
  Added:
    - Initial release, based on FactorioLab Export 1.0.7


control.lua
local dictionary = require("__flib__.dictionary")
local event = require("__flib__.event")

local export_data = require("scripts/export-data")

local function create_dictionaries()
  for _, type in pairs({
    "fluid", "item", "item_group", "recipe", "technology", "virtual_signal",
  }) do
    -- If the object's name doesn't have a translation, use its internal name as the translation
    local Names = dictionary.new(type .. "_names", true)
    for name, prototype in pairs(game[type .. "_prototypes"]) do
      Names:add(name, prototype.localised_name)
    end
  end
  local EntityNames = dictionary.new("entity_names", true)
  for name, prototype in pairs(game["entity_prototypes"]) do
    if game.noise_layer_prototypes["emoji-keyboard-export/entity/" .. name] ~= nil then
      EntityNames:add(name, prototype.localised_name)
    end
  end
  local OtherNames = dictionary.new("other_names", true)
  OtherNames:add("research", {"gui-technology-progress.title"})
  OtherNames:add("crafting", {"gui.crafting"})
end

event.on_init(
  function()
    dictionary.init()
    create_dictionaries()
  end
)

event.on_player_created(
  function(e)
    if game.noise_layer_prototypes["emoji-keyboard-export/"] == nil then
      -- Instrument mode is not enabled
      return
    end
    local player = game.players[e.player_index]
    player.print({"emoji-keyboard-export.initialize"})
    dictionary.translate(player)
  end
)

event.on_string_translated(
  function(e)
    local language_data = dictionary.process_translation(e)
    if language_data then
      for _, player_index in pairs(language_data.players) do
        export_data(player_index, language_data)
      end
    end
  end
)


scripts/json.lua
-- https://gist.github.com/tylerneylon/59f4bcf316be525b30ab

json = {}

local function kind_of(obj)
  if type(obj) ~= "table" then
    return type(obj)
  end

  local i = 1
  for _ in pairs(obj) do
    if obj[i] ~= nil then
      i = i + 1
    else
      return "table"
    end
  end

  if i == 1 then
    return "table"
  else
    return "array"
  end
end

local function escape_str(s)
  local in_char = {"\\", '"', "/", "\b", "\f", "\n", "\r", "\t"}
  local out_char = {"\\", '"', "/", "b", "f", "n", "r", "t"}
  for i, c in ipairs(in_char) do
    s = s:gsub(c, "\\" .. out_char[i])
  end

  return s
end

function json.stringify(obj, pretty, tabvalue, as_key)
  local tv = tabvalue
  if not tabvalue then
    tv = 0
  end

  local s = {} -- We'll build the string as an array of strings to be concatenated.
  local tab = pretty and "  " or ""
  local space = pretty and " " or ""
  local newline = pretty and "\n" or ""
  local kind = kind_of(obj) -- This is 'array' if it's an array or type(obj) otherwise.
  if kind == "array" then
    if as_key then
      error("Can't encode array as key.")
    end

    s[#s + 1] = "[" .. newline .. string.rep(tab, tv + 1)
    local t = {}
    for i, val in ipairs(obj) do
      if #t > 0 then
        t[#t + 1] = "," .. newline .. string.rep(tab, tv + 1)
      end

      t[#t + 1] = json.stringify(val, pretty, tv + 1)
    end
    s[#s + 1] = table.concat(t) .. newline .. string.rep(tab, tv) .. "]"
  elseif kind == "table" then
    if as_key then
      error("Can't encode table as key.")
    end

    s[#s + 1] = "{" .. newline .. string.rep(tab, tv + 1)
    local t = {}
    for k, v in pairs(obj) do
      if #t > 0 then
        t[#t + 1] = "," .. newline .. string.rep(tab, tv + 1)
      end

      t[#t + 1] = json.stringify(k, pretty, tv, true)
      t[#t + 1] = ":" .. space
      t[#t + 1] = json.stringify(v, pretty, tv + 1)
    end

    s[#s + 1] = table.concat(t) .. newline .. string.rep(tab, tv) .. "}"
  elseif kind == "string" then
    return '"' .. escape_str(obj) .. '"'
  elseif kind == "number" then
    if as_key or obj == -math.huge or obj == math.huge or obj ~= obj then
      return '"' .. tostring(obj) .. '"'
    end

    return tostring(obj)
  elseif kind == "boolean" then
    return tostring(obj)
  elseif kind == "nil" then
    return "null"
  else
    error("Unjsonifiable type: " .. kind .. ".")
  end

  return table.concat(s)
end

return json


scripts/utils.lua
utils = {}

-- Sort table by a list of fields
function utils.sort_by(t, n1, n2, n3, n4, n5, n6)
  local function sort_by_func(a, b)
    if a[n1] < b[n1] then
      return true
    elseif n2 and a[n1] == b[n1] then
      if a[n2] < b[n2] then
        return true
      elseif n3 and a[n2] == b[n2] then
        if a[n3] < b[n3] then
          return true
        elseif n4 and a[n3] == b[n3] then
          if a[n4] < b[n4] then
            return true
          elseif n5 and a[n4] == b[n4] then
            if a[n5] < b[n5] then
              return true
            elseif n6 and a[n5] == b[n5] then
              return a[n6] < b[n6]
            end
          end
        end
      end
    end

    return false
  end

  table.sort(t, sort_by_func)
end

-- Round a float to a number of decimals
function utils.round(num, decimals)
  local mult = 10 ^ (decimals or 0)
  return math.floor(num * mult + 0.5) / mult
end

-- Convert energy value to kW
function utils.convert_energy(energy)
  if not energy then
    return nil
  end

  return energy / 1000 * 60
end

-- Convert a list to a table by id/name
function utils.to_table(array)
  local table = {}
  for _, obj in pairs(array) do
    table[obj.id] = obj.name
  end

  return table
end

function utils.get_order_info(key)
  local noise_layer_name = "emoji-keyboard-export/" .. key
  local noise_layer = game.noise_layer_prototypes[noise_layer_name]
  if noise_layer == nil then
    return nil, nil
  end

  local hash_id, size = string.match(noise_layer.order, "^(.+)|(.+)$")
  if hash_id and size then
    return tonumber(hash_id), 64 / tonumber(size)
  else
    hash_id = string.match(noise_layer.order, "^(.+)$")
    if hash_id then
      return tonumber(hash_id), nil
    else
      return nil, nil
    end
  end
end

function utils.calculate_ingredients(ingredients)
  local lab_in = {}
  for _, ingredient in pairs(ingredients) do
    lab_in[ingredient.name] = ingredient.amount
  end

  return lab_in
end

function utils.calculate_products(products)
  local lab_out = {}
  local lab_catalyst
  local total = 0
  for _, product in pairs(products) do
    local amount = product.amount
    if not amount then
      amount = (product.amount_max + product.amount_min) / 2
    end

    if product.probability then
      amount = amount * product.probability
    end

    total = total + amount

    if lab_out[product.name] then
      lab_out[product.name] = lab_out[product.name] + amount
    else
    lab_out[product.name] = amount
    end

    if product.catalyst_amount then
      if not lab_catalyst then
        lab_catalyst = {}
      end

      lab_catalyst[product.name] = product.catalyst_amount
    end
  end

  return lab_out, lab_catalyst, total
end

return utils


scripts/collect-data.lua
local utils = require("utils")

local function process_technology()
  local recipes_unlocked = {}
  for _, technology in pairs(game.technology_prototypes) do
    local effects = technology.effects
    if effects then
      for _, effect in pairs(effects) do
        if effect and effect.type == "unlock-recipe" then
          recipes_unlocked[effect.recipe] = true
        end
      end
    end
  end

  return recipes_unlocked
end

local function process_recipes(recipes_unlocked)
  local recipes_enabled = {}
  local fixed_recipe = {}
  for _, entity in pairs(game.entity_prototypes) do
    if entity.fixed_recipe then
      fixed_recipe[entity.fixed_recipe] = true
    end
  end

  for name, recipe in pairs(game.recipe_prototypes) do
    local include = true

    -- Skip recipes that don't have outputs
    if not recipe.products or not (#recipe.products > 0) then
      include = false
    else
      local out, catalyst, total = utils.calculate_products(recipe.products)
      if total == 0 then
        include = false
      end
    end

    -- Always include fixed recipes (that do have outputs)
    if not fixed_recipe[name] then
      -- Skip recipes that are not unlocked / enabled
      if recipe.enabled == false and not recipes_unlocked[name] then
        include = false
      end

      -- Skip recipes that are hidden
      if recipe.hidden == true then
        include = false
      end
    end

    if include then
      recipes_enabled[name] = recipe
    end
  end
  return recipes_enabled
end

local function process_items(recipes_enabled)
  local items_used = {}

  -- Check for burnt result / rocket launch product
  for name, item in pairs(game.item_prototypes) do
    if #item.rocket_launch_products > 0 then
      items_used[name] = true
      for product_name, product in pairs(item.rocket_launch_products) do
        items_used[product.name] = true
      end
    end

    if item.burnt_result then
      items_used[name] = true
      items_used[item.burnt_result.name] = true
    end
  end

  -- Check for recipe input / output
  for name, recipe in pairs(recipes_enabled) do
    -- Check ingredients
    for _, ingredient in pairs(recipe.ingredients) do
      if ingredient then
        items_used[ingredient.name] = true
      end
    end

    -- Check products
    for _, product in pairs(recipe.products) do
      if product then
        items_used[product.name] = true
      end
    end
  end

  return items_used
end

local function sort_protos(items_used, recipes_enabled)
  local proto_list = {}
  for name, _ in pairs(items_used) do
    local item = game.item_prototypes[name]
    if item then
      table.insert(
        proto_list,
        {
          item = item,
          fluid = nil,
          recipe = nil,
          name = item.name,
          order = item.order,
          sgname = item.subgroup.name,
          sgorder = item.subgroup.order,
          gname = item.group.name,
          gorder = item.group.order
        }
      )
    end

    local fluid = game.fluid_prototypes[name]
    if fluid then
      table.insert(
        proto_list,
        {
          item = nil,
          fluid = fluid,
          recipe = nil,
          name = fluid.name,
          order = fluid.order,
          sgname = fluid.subgroup.name,
          sgorder = fluid.subgroup.order,
          gname = fluid.group.name,
          gorder = fluid.group.order
        }
      )
    end
  end

  for name, recipe in pairs(recipes_enabled) do
    table.insert(
      proto_list,
      {
        item = nil,
        fluid = nil,
        recipe = recipe,
        name = recipe.name,
        order = recipe.order,
        sgname = recipe.subgroup.name,
        sgorder = recipe.subgroup.order,
        gname = recipe.group.name,
        gorder = recipe.group.order
      }
    )
  end

  utils.sort_by(proto_list, "gorder", "gname", "sgorder", "sgname", "order", "name")

  local sorted_protos = {}
  for i = 1, #proto_list do
    local proto = proto_list[i]
    table.insert(
      sorted_protos,
      {
        item = proto.item,
        fluid = proto.fluid,
        recipe = proto.recipe
      }
    )
  end

  return sorted_protos
end

return function()
  local recipes_unlocked = process_technology()
  local recipes_enabled = process_recipes(recipes_unlocked)
  local items_used = process_items(recipes_enabled)
  return sort_protos(items_used, recipes_enabled)
end


instrument-after-data.lua
local noise_layers = {} -- Use fake noise layers to store sprite info
local prefix = "emoji-keyboard-export/"

table.insert(noise_layers, {name = prefix, order = prefix, type = "noise-layer"})

local icon_hash, icon_hash_id = {}, 0
local function get_icon_hash(obj)
  if not obj.icon and not obj.icons then
    return -1
  end

  local serialized = string.format("%s.%s.%s", obj.icon or "", obj.icon_size or "", obj.icon_mipmaps or "")
  if obj.icons then
    for _, icon in pairs(obj.icons) do
      serialized =
        string.format(
        "%s,%s.%s.%s.%s.%s.%s.%s.%s.%s.%s",
        serialized,
        icon.icon,
        icon.icon_size,
        icon.scale or "",
        icon.icon_mipmaps or "",
        (icon.shift and icon.shift[1]) or "",
        (icon.shift and icon.shift[2]) or "",
        (icon.tint and (icon.tint.r or icon.tint[1])) or "",
        (icon.tint and (icon.tint.g or icon.tint[2])) or "",
        (icon.tint and (icon.tint.b or icon.tint[3])) or "",
        (icon.tint and (icon.tint.a or icon.tint[4])) or ""
      )
    end
  end
  if not icon_hash[serialized] then
    icon_hash[serialized] = icon_hash_id
    icon_hash_id = icon_hash_id + 1
  end
  return icon_hash[serialized]
end

local function check_overflow(obj)
  if not obj.icons or #obj.icons < 2 then
    return nil
  end
  local base = obj.icons[1]
  local size = (base.icon_size or obj.icon_size) * (base.scale or 1)
  if base.icon_size == 64 and base.scale == nil then
    size = 32
  end
  local compare_size = size

  for _, icon in pairs(obj.icons) do
    if icon.shift then
      local ico_size = (icon.icon_size or obj.icon_size) * (icon.scale or 1)
      if icon.icon_size == 64 and icon.scale == nil then
        ico_size = 32
      end
      local a = (icon.shift[1] < 0 and icon.shift[1] * -1) or icon.shift[1]
      local b = (icon.shift[2] < 0 and icon.shift[2] * -1) or icon.shift[2]
      local offset = (a > b and a) or b
      local width = ((ico_size / 2) + offset) * 2

      if width > compare_size then
        compare_size = (width / size) * 32
      end
    end
  end

  if compare_size ~= size then
    -- log(obj.name .. ": " .. compare_size)
    -- for _, i in pairs(obj.icons) do
    --   log(
    --     "icon: " ..
    --       i.icon ..
    --         ", mips: " ..
    --           (i.icon_mipmaps or "nil") ..
    --             ", size: " ..
    --               (i.icon_size or "nil") ..
    --                 ", scale: " ..
    --                   (i.scale or "nil") .. ", shift: " .. ((i.shift and (i.shift[1] .. "|" .. i.shift[2])) or "nil")
    --   )
    -- end
    return compare_size
  else
    return nil
  end
end

local function check_size(key, store_as)
  local raw = data.raw[key]
  for _, obj in pairs(raw) do
    local order = get_icon_hash(obj)
    local size = check_overflow(obj)
    if not size and obj.icons and obj.icons[1].scale == 1 then
      size = obj.icons[1].icon_size
    end
    if size then
      order = order .. "|" .. size
    end
    local name = prefix .. (store_as or key) .. "/" .. obj.name
    table.insert(noise_layers, {name = name, order = order, type = "noise-layer"})
  end
end

local function check_size_alt(key)
  local raw = data.raw[key]
  for _, obj in pairs(raw) do
    local order = get_icon_hash(obj)
    if obj.icons and obj.icons[1].icon_size then
      order = order .. "|" .. (obj.icons[1].icon_size * (obj.icons[1].scale or 1))
    else
      order = order .. "|" .. obj.icon_size
    end
    local name = prefix .. key .. "/" .. obj.name
    table.insert(noise_layers, {name = name, order = order, type = "noise-layer"})
  end
end

check_size("item")
check_size("ammo", "item")
check_size("armor", "item")
check_size("capsule", "item")
check_size("gun", "item")
check_size("item-with-entity-data", "item")
check_size("module", "item")
check_size("rail-planner", "item")
check_size("repair-tool", "item")
check_size("spidertron-remote", "item")
check_size("tool", "item")
check_size("fluid")
check_size("recipe")
check_size("virtual-signal")
check_size("character", "entity")
check_size("cliff", "entity")
check_size("corpse", "entity")
check_size("entity-ghost", "entity")
check_size("item-entity", "entity")
check_size("item-request-proxy", "entity")
check_size("tile-ghost", "entity")
check_size("tree", "entity")
check_size("turret", "entity")
check_size("unit", "entity")
check_size("unit-spawner", "entity")
check_size_alt("item-group")
check_size_alt("technology")

data:extend(noise_layers)


scripts/export-data.lua
local collect_data = require("collect-data")
local json = require("json")
local utils = require("utils")

local folder = "emoji-keyboard-export/"
local color_error = {r = 1, g = 0, b = 0}
local color_warn = {r = 1, g = 0.5, b = 0}
local color_good = {r = 0, g = 1, b = 0}
local player

local function add_icon(hash_id, name, scale, sprite, icons)
  if hash_id ~= -1 then
    if hash_id then
      for _, icon in pairs(icons) do
        if icon.hash_id == hash_id then
          table.insert(icon.copies, name)
          return
        end
      end
    end

    table.insert(icons, {hash_id = hash_id, name = name, scale = scale, sprite = sprite, copies = {}})
  end
end

local function check_recipe_name(recipe_id_claimed, desired_id, backup_id, icons)
  if recipe_id_claimed[desired_id] then
    for _, icon in pairs(icons) do
      if icon.name == desired_id then
        table.insert(icon.copies, backup_id)
        break
      end
    end

    recipe_id_claimed[backup_id] = true
    return backup_id
  end

  recipe_id_claimed[desired_id] = true
  return desired_id
end

local function check_icon_name(desired_id, backup_id, icons)
  for _, icon in pairs(icons) do
    if icon.name == desired_id then
      return backup_id
    else
      for _, copy in pairs(icon.copies) do
        if copy == desired_id then
          return backup_id
        end
      end
    end
  end

  return desired_id
end

return function(player_index, language_data)
  player = game.players[player_index]
  local player_settings = settings.get_player_settings(player)
  local dictionaries = language_data.dictionaries
  local language = language_data.language

  local function warn_player(msg)
    player.print(msg, color_warn)
  end

  -- Localized names
  local group_names = dictionaries["item_group_names"]
  local item_names = dictionaries["item_names"]
  local fluid_names = dictionaries["fluid_names"]
  local recipe_names = dictionaries["recipe_names"]
  local technology_names = dictionaries["technology_names"]
  local gui_names = dictionaries["other_names"]
  local virtual_signal_names = dictionaries["virtual_signal_names"]
  local entity_names = dictionaries["entity_names"]

  local sorted_protos = collect_data()
  local icons = {}
  local recipe_id_claimed = {}

  -- Final data collections
  local sprite_map_index = {}
  local by_group = {}
  local by_group_recipes = {}

  local board_name = "Factorio"
  local board_sprite = "inserter"

  -- Build ID based on active mods
  local build_id = game.active_mods['base'] .. "-"
  local version = {}
  if game.active_mods['space-exploration'] ~= nil then
    build_id = build_id .. "se" .. game.active_mods['space-exploration'] .. "-"
    board_name = board_name .. " SE"
    board_sprite = "se-spaceship"
  end
  for name, ver in pairs(game.active_mods) do
    if name ~= "emoji-keyboard-export" and name ~= "flib" then
      version[name] = ver
    end
  end
  build_id = build_id .. language
  local sprite_map_name = "factorio" .. "-" .. build_id
  local sprite_map_path = "factorio-icons-" .. build_id .. ".png"
  local plugin_file_path = "factorio-plugin-" .. build_id .. ".json"

  group_names["research"] = gui_names["research"]
  for name, _ in pairs(group_names) do
    by_group[name] = {}
    by_group_recipes[name] = {}
  end

  -- Process items
  for _, proto in pairs(sorted_protos) do
    if proto.item then
      local item = proto.item
      local name = item.name
      table.insert(by_group[item.group.name], {
        cluster = "[item=" .. name .. "]",
        name = item_names[name],
        symbol = {
          spriteMap = sprite_map_name,
          sprite = name,
        },
      })
      local hash_id, scale = utils.get_order_info("item/" .. name)
      add_icon(hash_id, name, scale or 2, "item/" .. name, icons)
    end

    if proto.fluid then
      local fluid = proto.fluid
      local name = fluid.name
      if fluid then
        table.insert(by_group[fluid.group.name], {
          cluster = "[fluid=" .. name .. "]",
          name = fluid_names[name],
          symbol = {
            spriteMap = sprite_map_name,
            sprite = name,
          },
        })
        local hash_id, scale = utils.get_order_info("fluid/" .. name)
        add_icon(hash_id, name, scale or 2, "fluid/" .. name, icons)
      else
        player.print({"emoji-keyboard-export.error-no-item-prototype", name}, color_error)
      end
    end

    if proto.recipe then
      recipe_id_claimed[proto.recipe.name] = true
    end
  end

  -- Process recipes
  for _, proto in pairs(sorted_protos) do
    if proto.recipe then
      local recipe = proto.recipe
      local name = recipe.name
      local hash_id, scale = utils.get_order_info("recipe/" .. name)
      if hash_id and hash_id ~= -1 then
        local icon_id = check_icon_name(name, name .. "|recipe", icons)

        table.insert(by_group_recipes[recipe.group.name], {
          cluster = "[recipe=" .. name .. "]",
          name = recipe_names[name],
          symbol = {
            spriteMap = sprite_map_name,
            sprite = icon_id,
          },
        })

        add_icon(hash_id, icon_id, scale or 2, "recipe/" .. name, icons)
      end
    end
  end

  -- Process signals
  for name, _signal in pairs(game.virtual_signal_prototypes) do
    local hash_id, scale = utils.get_order_info("virtual-signal/" .. name)
    local icon_id = check_icon_name(name, name .. "|signal", icons)
    table.insert(by_group["signals"], {
      cluster = "[virtual-signal=" .. name .. "]",
      name = virtual_signal_names[name],
      symbol = {
        spriteMap = sprite_map_name,
        sprite = icon_id,
      },
    })

    add_icon(hash_id, icon_id, scale or 2, "virtual-signal/" .. name, icons)
  end

  -- Process entities
  for name, entity in pairs(game.entity_prototypes) do
    local remnants = "-remnants"
    local hash_id, scale = utils.get_order_info("entity/" .. name)
    if hash_id ~= nil and name:sub(-#remnants) ~= remnants then -- and entity.selectable_in_game 
      local icon_id = check_icon_name(name, name .. "|entity", icons)
      table.insert(by_group[entity.group.name], {
        cluster = "[entity=" .. name .. "]",
        name = entity_names[name],
        symbol = {
          spriteMap = sprite_map_name,
          sprite = icon_id,
        },
      })
      add_icon(hash_id, icon_id, scale or 2, "entity/" .. name, icons)
    end
  end

  -- Process Technologies
  for name, tech in pairs(game.technology_prototypes) do
    local desired_id = name
    local backup_id = name .. "-technology"
    local id = check_recipe_name(recipe_id_claimed, desired_id, backup_id, {})
    local hash_id, scale = utils.get_order_info("technology/" .. name)
    add_icon(hash_id, id, scale or 0.25, "technology/" .. name, icons)
    local icon_id = id

    table.insert(by_group["research"], {
      cluster = "[technology=" .. name .. "]",
      name = technology_names[name],
      symbol = {
        spriteMap = sprite_map_name,
        sprite = icon_id,
      },
    })
  end

  -- Process categories
  local boards = {}
  local recipes_boards = {}
  for name, _ in pairs(by_group) do
    local icon_id
    if name == "research" then
      icon_id = "research"
      local sprite = "space-science-pack"
      if game.active_mods["space-exploration"] then
        sprite = "se-rocket-science-pack"
      end
    
      if not game.technology_prototypes[sprite] then
        player.print({"emoji-keyboard-export.error-no-research-sprite", sprite}, color_error)
      else
        sprite = "technology/" .. sprite
        local hash_id, scale = utils.get_order_info(sprite)
        add_icon(hash_id, icon_id, scale or 0.25, sprite, icons)
      end
    else
      local hash_id, scale = utils.get_order_info("item-group/" .. name)
      icon_id = check_icon_name(name, name .. "|category", icons)
      add_icon(hash_id, icon_id, scale or 0.25, "item-group/" .. name, icons)
    end

    if #by_group[name] > 0 then
      table.insert(boards, {
        name = group_names[name],
        symbol = {
          spriteMap = sprite_map_name,
          sprite = icon_id,
        },
        content = by_group[name],
      })
    end
    if #by_group_recipes[name] > 0 then
      table.insert(recipes_boards, {
        name = group_names[name] .. "\n" .. gui_names["crafting"],
        symbol = {
          spriteMap = sprite_map_name,
          sprite = icon_id,
        },
        content = by_group_recipes[name],
      })
    end
  end

  game.remove_path(folder)
  local pretty_json = true --player_settings["emoji-keyboard-export-pretty-json"].value

  -- Process and generate sprite for scaled icons
  local sprite_surface = game.create_surface("lab-sprite")

  -- Calculate sprite sheet width (height determined by # of loop iterations)
  local width = math.ceil((#icons) ^ 0.5)
  if width < 8 then
    width = 8
  end

  local tile_size = 32
  local sprite_size = 64 -- each sprite takes 2×2 tiles once scaled
  local sprite_pad = 8
  local stride = (sprite_size + sprite_pad) / tile_size
  local x_position = stride / 2 * (width - 1)
  local x_resolution = width * (sprite_size + sprite_pad)

  local x = 0
  local y = 0
  for _, icon in pairs(icons) do
    rendering.draw_sprite(
      {
        sprite = icon.sprite,
        x_scale = icon.scale,
        y_scale = icon.scale,
        target = {x = x * stride, y = y * stride},
        surface = sprite_surface
      }
    )
    local position = {
      row = y,
      col = x
    }
    sprite_map_index[icon.name] = position
    for _, copy in pairs(icon.copies) do
      sprite_map_index[copy] = position
    end

    x = x + 1
    if x == width then
      y = y + 1
      x = 0
    end
  end

  if x == 0 then
    y = y - 1
  end

  local rows = y + 1
  local y_resolution = rows * (sprite_size + sprite_pad)
  local y_position = stride / 2 * (rows - 1)

  game.take_screenshot(
    {
      player = player,
      by_player = player,
      surface = sprite_surface,
      position = {x_position, y_position},
      resolution = {x_resolution, y_resolution},
      zoom = 1,
      quality = 100,
      daytime = 1,
      path = folder .. sprite_map_path,
      show_gui = false,
      show_entity_info = false,
      anti_alias = false
    }
  )

  local sprite_maps = {}
  sprite_maps[sprite_map_name] = {
    path = sprite_map_path,
    width = sprite_size,
    height = sprite_size,
    padding = sprite_pad / 2,
    cols = width,
    rows = rows,
    index = sprite_map_index,
  }

  local lab_data = {
    name = board_name,
    symbol = {
      spriteMap = sprite_map_name,
      sprite = board_sprite,
    },
    boards = {{
      name = board_name,
      symbol = {
        spriteMap = sprite_map_name,
        sprite = board_sprite,
      },
      byRow = {
        {},
        boards,
        recipes_boards,
      }
    }},
    spriteMaps = sprite_maps,
    __version = version,
    __language = language,
  }

  game.write_file(folder .. plugin_file_path, json.stringify(lab_data, pretty_json))
  player.print({"emoji-keyboard-export.complete-data"}, color_good)
end
